crates/hir-ty/src/infer.rs

```diff
@@ -67,6 +67,9 @@ pub(crate) fn infer_query(db: &dyn HirDatabase, def: DefWithBodyId) -> Arc<Infer
         DefWithBodyId::ConstId(c) => ctx.collect_const(&db.const_data(c)),
         DefWithBodyId::FunctionId(f) => ctx.collect_fn(f),
         DefWithBodyId::StaticId(s) => ctx.collect_static(&db.static_data(s)),
+        DefWithBodyId::VariantId(v) => {
+            ctx.return_ty = TyBuilder::def_ty(db, v.parent.into()).fill_with_unknown().build()
```

Comment:
I thought so too initially, but this actually returns the type of the variant, which is actually the type of the enum itself.

Reply:
So in `enum E { A }`, you have E::A being of the type E, while if this returned the integer inside the E::A would be some integer

Reply:
Huh, I don't see how that snippet is relevant here. The infer query is what runs on the variants expressions no? That is for
```rs
enum E {
    V = 3
}
```
The body here is the `3` is it not? So this should definitely expect an integer type, not the enum.

Reply:
So when I tried implementing it like that it broke stuff like `E::A as u8`, because E::A got inferred to be an integer, but that's not valid. You can't get the number inside an enum without casting it explicitly

Reply:
If I change this to just assume `usize` as the return type the const eval test works fine for me though, I feel like you are mixing up what the return type here means? The return type is strictly for what the expression in the variant has to resolve to, not for what `E::V` resolves to when used, and judging from how you have implemented the `cast` expression in const eval for enums, you are already using it properly with assuming the `lit` (which is what you have calculated in the enum variants body, this has to be in integer otherwise the cast wouldn't cast anything). Point is, if you assume that in
```rs
enum E {
    V = 3
}
```
The `3` expression has to be of type `E`, we'll get type mismatches (once we diagnose these bodies) since we clearly want to expect an integer here, not an `E`

Reply:
I had another look at the code, and I now agree with you, I'm calculating the value of the specific variant, and then wrapping it as an enum afterwards to force a cast to be necessary, but this value refers to the result of `eval_const_variant`, which is the internal value of the variant.
Thanks!